google/cloud/completion_queue.h

```diff
@@ -31,8 +31,19 @@ GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
 class CompletionQueue;
 
 namespace internal {
+
 std::shared_ptr<CompletionQueueImpl> GetCompletionQueueImpl(
     CompletionQueue const& cq);
+
+template <
+    typename AsyncCallType, typename Request,
```

Comment:
There may be an opportunity to simplify this.

https://godbolt.org/z/dhGqTWsbv

Where "simplify" means "use less hand-crafted type deduction and let the compiler do it".

Reply:
TIL: absl includes in godbolt.

A consequence of this seems to be that we have to explicitly name the `Request` and `Response` types when calling the method. Does that seem right?

```
<lambda(grpc::ClientContext*, const google::bigtable::admin::v2::GetTableRequest&, grpc::CompletionQueue*)>' 
is not derived from
'absl::lts_20230125::FunctionRef<std::unique_ptr<grpc::ClientAsyncResponseReaderInterface<Response> >(grpc::ClientContext*, const Request&, grpc::CompletionQueue*)>'
```

Reply:
> 
> A consequence of this seems to be that we have to explicitly name the `Request` and `Response` types when calling the method. Does that seem right?

That seems unfortunate, but for an internal / implementation function it may be a plus.  Your call, I am not convinced my idea improved things.  But maybe it did.

Reply:
Let's go with it. If you ask me, it is easier to understand.